contracts-bedrock: cleanup interop predeploys

tynes opened the pull request:
**Description**

Small refactor to the interop predeploys to reuse existing code.
It was determined safe to use the `SafeCall` library, so we opt
to use that instead of duplicating code. No tests are added since
existing test coverage ensures that these calls happen as expected.
A manual mutation test where the value was not passed through was
performed and failing tests were observed.

The `SafeCall` library was previously pinned to `0.8.15` due to the
"call with min gas" semantics, it may be possible that a new compiler
version could emit code that doesn't act how `0.8.15` acts, but the
whole gas introspection thing is an anti pattern and we have invariant
tests that would catch this if we modified the solc version used
by `OptimismPortal`. We know to never follow this pattern again,
ie `hasMinGas` or `callWithMinGas` should never be used again
in the codebase and should be deleted at some point once we get
rid of the whole min gas limit semantics.

<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

tynes committed:
contracts-bedrock: cleanup interop predeploys

Small refactor to the interop predeploys to reuse existing code.
It was determined safe to use the `SafeCall` library, so we opt
to use that instead of duplicating code. No tests are added since
existing test coverage ensures that these calls happen as expected.
A manual mutation test where the value was not passed through was
performed and failing tests were observed.

The `SafeCall` library was previously pinned to `0.8.15` due to the
"call with min gas" semantics, it may be possible that a new compiler
version could emit code that doesn't act how `0.8.15` acts, but the
whole gas introspection thing is an anti pattern and we have invariant
tests that would catch this if we modified the solc version used
by `OptimismPortal`. We know to never follow this pattern again,
ie `hasMinGas` or `callWithMinGas` should never be used again
in the codebase and should be deleted at some point once we get
rid of the whole min gas limit semantics.

tynes committed:
lint: fix

tynes committed:
semver-lock: update

mds1 commented:
>  it may be possible that a new compiler version could emit code that doesn't act how `0.8.15` acts, but the whole gas introspection thing is an anti pattern and we have invariant tests that would catch this if we modified the solc version used by `OptimismPortal`

Can we do a manual mutation test to make sure the invariant tests would actually catch this, e.g. just burn some extra gas in between and make sure it fails? I want to verify because it seems at least 1 of our invariant tests might not actually work properly, ref https://github.com/ethereum-optimism/client-pod/issues/625 where it seemed the ResourceMetering invariant test did not catch a bug

tynes commented:
> > it may be possible that a new compiler version could emit code that doesn't act how `0.8.15` acts, but the whole gas introspection thing is an anti pattern and we have invariant tests that would catch this if we modified the solc version used by `OptimismPortal`
> 
> Can we do a manual mutation test to make sure the invariant tests would actually catch this, e.g. just burn some extra gas in between and make sure it fails? I want to verify because it seems at least 1 of our invariant tests might not actually work properly, ref [ethereum-optimism/client-pod#625](https://github.com/ethereum-optimism/client-pod/issues/625) where it seemed the ResourceMetering invariant test did not catch a bug



> > it may be possible that a new compiler version could emit code that doesn't act how `0.8.15` acts, but the whole gas introspection thing is an anti pattern and we have invariant tests that would catch this if we modified the solc version used by `OptimismPortal`
> 
> Can we do a manual mutation test to make sure the invariant tests would actually catch this, e.g. just burn some extra gas in between and make sure it fails? I want to verify because it seems at least 1 of our invariant tests might not actually work properly, ref [ethereum-optimism/client-pod#625](https://github.com/ethereum-optimism/client-pod/issues/625) where it seemed the ResourceMetering invariant test did not catch a bug

This is only relevant if the `OptimismPortal` has its solc version changed. This doesn't do that, so I am wondering what specific thing you are asking for

mds1 commented:
I read your message as saying "Here we loosen the SafeCall pragma, so if we bump OptimismPortal's solc version in the future, it's possible that we break the gas stipend invariants. This PR's pragma loosen is justified because our invariant tests would catch the future portal change"

So I was asking about proactively verifying the "our invariant tests would catch that" claim on this PR. There is an argument that it's out of scope since we'll need to remember to verify that invariant anyway, though.

tynes commented:
> I read your message as saying "Here we loosen the SafeCall pragma, so if we bump OptimismPortal's solc version in the future, it's possible that we break the gas stipend invariants. This PR's pragma loosen is justified because our invariant tests would catch the future portal change"
> 
> So I was asking about proactively verifying the "our invariant tests would catch that" claim on this PR. There is an argument that it's out of scope since we'll need to remember to verify that invariant anyway, though.

I'd like to bucket this work in the eventual "bump to modern solc" project if possible as this is tangential to this PR

mds1 commented:
That works for me, thanks

tynes commented:
Thank you @mds1 !
